src/PreparedWrites.h

```diff
@@ -0,0 +1,227 @@
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef RAMCLOUD_PREPAREDWRITES_H
+#define RAMCLOUD_PREPAREDWRITES_H
+
+#include <map>
+#include <utility>
+#include "Common.h"
+#include "Object.h"
+#include "SpinLock.h"
+#include "WorkerTimer.h"
+
+namespace RAMCloud {
+
+class ObjectManager;
+
+class PreparedOp {
+  public:
+    PreparedOp(WireFormat::TxPrepare::OpType type,
+               uint64_t clientId, uint64_t rpcId,
+               uint32_t participantCount,
+               WireFormat::TxParticipant* participants,
+               uint64_t tableId, uint64_t version, uint32_t timestamp,
+               Buffer& keysAndValueBuffer, uint32_t startDataOffset = 0,
+               uint32_t length = 0);
+    PreparedOp(WireFormat::TxPrepare::OpType type,
+               uint64_t clientId, uint64_t rpcId,
+               uint32_t participantCount,
+               WireFormat::TxParticipant* participants,
+               Key& key, const void* value, uint32_t valueLength,
+               uint64_t version, uint32_t timestamp,
+               Buffer& buffer, uint32_t *length = NULL);
+    PreparedOp(Buffer& buffer, uint32_t offset, uint32_t length);
+
+    class Header {
+      public:
+        Header(WireFormat::TxPrepare::OpType type,
+               uint64_t clientId,
+               uint64_t rpcId,
+               uint32_t participantCount)
+            : type(type),
+              clientId(clientId),
+              rpcId(rpcId),
+              participantCount(participantCount),
+              checksum(0)
+        {
+        }
+
+        /// Type of the staged operation.
+        WireFormat::TxPrepare::OpType type;
+
+        /// leaseId given for this prepare.
+        uint64_t clientId;
+
+        /// rpcId given for this prepare.
+        uint64_t rpcId;
+
+        /// Number of objects participating current transaction.
+        uint32_t participantCount;
+
+        /// CRC32C checksum covering everything but this field, including the
+        /// keys and the value.
+        uint32_t checksum;
+
+    } __attribute__((__packed__));
+    static_assert(sizeof(Header) == 28,
+        "Unexpected serialized PreparedOp size");
+
+    /// Copy of the PreparedOp header.
+    Header header;
+
+    /// Pointer to the array of #WireFormat::TxParticipant which contains
+    /// all information of objects participating transaction.
+    WireFormat::TxParticipant* participants;
+
+    /// Object to be written during COMMIT phase.
+    /// Its value is empty if OpType is READ or REMOVE; it only contains
+    /// key information in that case.
+    Object object;
+
+    void assembleForLog(Buffer& buffer);
+    bool checkIntegrity();
+    uint32_t computeChecksum();
+
+    DISALLOW_COPY_AND_ASSIGN(PreparedOp);
+};
+
+class PreparedOpTombstone {
```

Comment:
Need overall class documentation. I'm not sure why this information has been consistently left out; let's discuss.